smaug/python/node.py

```diff
@@ -0,0 +1,125 @@
+import sys
+import numpy as np
+
+from smaug.core import node_pb2
+from smaug.core import types_pb2
+from smaug.python import global_vars
+from smaug.python import datatypes
+
+class Node:
+  def __init__(self, name, op, params=None, inputs=None, outputs=None):
+    """Create a node.
+
+    A `Node` instance contains information about its corresponding operation,
+    including the operator type, parameters and input/output tensors. A `Graph`
+    is made up of `Node`s. When serialized, a `NodeProto` is created.
+
+    Args:
+      name: Name of the node.
+      op: `OpType` representing the operation type of the node.
+      params: `Params` used by the operator (optional).
+      inputs: A list of `Tensor` (optional).
+      outputs: A list of `Tensor` (optional).
+
+    Returns:
+      A `Node` instance.
+    """
+    self._name = name
+    self._op = op
+    self._params = params
+    self._inputs = [] if inputs is None else inputs
+    self._outputs = [] if outputs is None else outputs
+
+  @property
+  def name(self):
+    return self._name
+
+  @property
+  def op(self):
+    return self._op
+
+  @property
+  def inputs(self):
+    return self._inputs
+
+  @property
+  def outputs(self):
+    return self._outputs
+
+  def add_input(self, tensor):
+    """Add an input tensor to the node.
+
+    Args:
+      tensor: A `Tensor`.
+    """
+    self._inputs.append(tensor)
+
+  def add_output(self, tensor):
+    """Add an output tensor to the node.
+
+    Args:
+      tensor: A `Tensor`.
+    """
+    self._outputs.append(tensor)
+
+  def update_input(self, tensor, index):
+    """Update the `index`th input with `tensor`.
+
+    Args:
+      tensor: A `Tensor` representing the new input.
+      index: The input index.
+    """
+    try:
+      self._inputs[index] = tensor
+    except IndexError:
```

Comment:
Sorry I wasn't clear on this. Yes, don't exit here, but also, don't just swallow the exception. This is not an exception you are expecting to handle; it's clearly a bug that needs to be fixed, so you should just not catch the exception at all.

Reply:
I see. Done.